### PR TITLE
add support for 0 sized matrices in svd

### LINKED ISSUE #2371: Incorrect behavior with empty arrays for `linalg.qr` and `linalg.svd`
`qr` and `svd` routines raise `CUSOLVERError: CUSOLVER_STATUS_INVALID_VALUE` for zero-sized input arrays.
NumPy supports this kind of input for `qr`, `svd` since 1.16
https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#linalglstsq-linalgqr-and-linalgsvd-now-work-with-empty-arrays
The problem is that when calling cuSOLVER routines leading dimension is passed as `input.shape[0]` but it should be `max(1, input.shape[0])`.

### PR DESCRIPTION
**Pull request by rushabh-v**
closes #2371

### CODE REVIEW

**Comment by rushabh-v** [comment]
I have tested it. The tests are passing.

File: cupy/linalg/decomposition.py [hunk 1]
```diff
@@ -357,7 +357,16 @@ def svd(a, full_matrices=True, compute_uv=True):
...
         x = a.transpose().astype(a_dtype, order='C', copy=True)
         trans_flag = True
+
     mn = min(m, n)
+    if mn == 0:
+        if compute_uv:
+            if full_matrices:
+                return cupy.eye(m, m), cupy.array([]), cupy.empty((0, 0))
+            else:
+                return cupy.empty((m, 0)), cupy.array([]), cupy.empty((0, 0))
```
Comment (asi1024) [inline cupy/linalg/decomposition.py]: This code path looks not tested.

File: cupy/linalg/decomposition.py [hunk 2]
```diff
@@ -357,7 +357,16 @@ def svd(a, full_matrices=True, compute_uv=True):
...
+        else:
+            return cupy.array([])
```
Comment (asi1024) [inline cupy/linalg/decomposition.py]: Ditto.

File: cupy/linalg/decomposition.py [hunk 3]
```diff
@@ -357,7 +357,16 @@ def svd(a, full_matrices=True, compute_uv=True):
         m, n = a.shape
         x = a.transpose().astype(a_dtype, order='C', copy=True)
         trans_flag = True
+
     mn = min(m, n)
+    if mn == 0:
+        if compute_uv:
+            if full_matrices:
+                return cupy.eye(m, m), cupy.array([]), cupy.empty((0, 0))
```
Comment (asi1024) [inline cupy/linalg/decomposition.py]: How about writing as follows for the consistency?
```suggestion
                return cupy.eye(m, m), cupy.empty((0,)), cupy.empty((0, 0))
```

File: cupy/linalg/decomposition.py (hunk 1 above)
Comment (rushabh-v) [reply cupy/linalg/decomposition.py]: wouldn't it be directly tested by the testing.product as it's `'full_matrices': [True, False]` [here](https://github.com/cupy/cupy/blob/86eab8f0ebc124d9e6b746efa0d635d6f37062ae/tests/cupy_tests/linalg_tests/test_decomposition.py#L116) and `self.full_metrices` has been passed [here](https://github.com/cupy/cupy/blob/86eab8f0ebc124d9e6b746efa0d635d6f37062ae/tests/cupy_tests/linalg_tests/test_decomposition.py#L135)?

File: cupy/linalg/decomposition.py (hunk 2 above)
Comment (rushabh-v) [reply cupy/linalg/decomposition.py]: I added that case now.

File: cupy/linalg/decomposition.py (hunk 3 above)
Comment (rushabh-v) [reply cupy/linalg/decomposition.py]: Done

File: tests/cupy_tests/linalg_tests/test_decomposition.py [hunk 4]
```diff
@@ -203,3 +203,16 @@ def test_svd_no_uv(self):
...
         self.check_rank2(cupy.random.randn(1, 2, 3, 4).astype(numpy.float64))
+
+    @testing.with_requires('numpy>=1.16')
+    def test_empty_array(self):
+        self.check_usv((0, 3))
+        self.check_usv((3, 0))
+        self.check_usv((1, 0))
+
+    @testing.with_requires('numpy>=1.16')
+    @testing.numpy_cupy_allclose(atol=1e-4)
```
Comment (asi1024) [inline tests/cupy_tests/linalg_tests/test_decomposition.py]: In this test, the result arrays can be compared without tolerances.
```suggestion
    @testing.numpy_cupy_array_equal()
```

File: cupy/linalg/decomposition.py (hunk 1 above)
Comment (asi1024) [reply cupy/linalg/decomposition.py]: I confirmed that this line is covered. Thanks!

File: tests/cupy_tests/linalg_tests/test_decomposition.py (hunk 4 above)
Comment (rushabh-v) [reply tests/cupy_tests/linalg_tests/test_decomposition.py]: I have made that change.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM!

**Pull request merged**